c++/src/capnp/serialize-async.h

```diff
@@ -76,6 +76,107 @@ kj::Promise<void> writeMessage(kj::AsyncCapabilityStream& output, kj::ArrayPtr<c
     KJ_WARN_UNUSED_RESULT;
 // Write a message with FDs attached, e.g. to a Unix socket with SCM_RIGHTS.
 
+// -----------------------------------------------------------------------------
+// Versions virtualized over an interface.
+
+class MessageTransport {
+  // Interface over which messages can be sent and received; virtualizes
+  // the functionality above.
+  //
+  // The methods exposed by this class have the same contracts as the corresponding
+  // FD-oriented functions above, except that the first argument (of type
+  // AsyncCapabilityStream) is omitted.
+  //
+  // Only the most general variants are exposed here; callers can use the
+  // functions below for the other variants.
+  //
+  // Implementations which do not support FD passing may simply ignore the relevant
+  // arguments.
+public:
+  virtual kj::Promise<kj::Maybe<MessageReaderAndFds>> tryReadMessage(
+      kj::ArrayPtr<kj::AutoCloseFd> fdSpace,
+      ReaderOptions options = ReaderOptions(), kj::ArrayPtr<word> scratchSpace = nullptr) = 0;
+
+  virtual kj::Promise<void> writeMessage(
+      kj::ArrayPtr<const int> fds,
+      kj::ArrayPtr<const kj::ArrayPtr<const word>> segments) = 0;
+    KJ_WARN_UNUSED_RESULT;
+};
+
+class AsyncIoTransport: public MessageTransport {
+  // A MessageTransport that wraps an AsyncIoStream.
+public:
+  explicit AsyncIoTransport(kj::AsyncIoStream& stream);
+
+  // Implements MessageTransport
+  kj::Promise<kj::Maybe<MessageReaderAndFds>> tryReadMessage(
+      kj::ArrayPtr<kj::AutoCloseFd> fdSpace,
+      ReaderOptions options = ReaderOptions(), kj::ArrayPtr<word> scratchSpace = nullptr) override;
+  kj::Promise<void> writeMessage(
+      kj::ArrayPtr<const int> fds,
+      kj::ArrayPtr<const kj::ArrayPtr<const word>> segments) override;
+private:
+  kj::AsyncIoStream& stream;
+};
+
+class AsyncCapabilityTransport: public MessageTransport {
+  // A MessageTransport that wraps an AsyncCapabilityStream.
+public:
+  explicit AsyncCapabilityTransport(kj::AsyncCapabilityStream& stream);
+
+  // Implements MessageTransport
+  kj::Promise<kj::Maybe<MessageReaderAndFds>> tryReadMessage(
+      kj::ArrayPtr<kj::AutoCloseFd> fdSpace,
+      ReaderOptions options = ReaderOptions(), kj::ArrayPtr<word> scratchSpace = nullptr) override;
+  kj::Promise<void> writeMessage(
+      kj::ArrayPtr<const int> fds,
+      kj::ArrayPtr<const kj::ArrayPtr<const word>> segments) override;
+private:
+  kj::AsyncCapabilityStream& stream;
+};
+
+kj::Promise<kj::Own<MessageReader>> readMessage(
```

Comment:
I think these freestanding functions should be methods of `MessageTransport` instead. This is consistent with e.g. `AsyncInputStream` which has `read()` defined as a method on the base class, which calls `tryRead()` and converts EOF into an exception.

The reason the existing functions in this file are free-standing is because they don't really belong on any type in particular, but `MessageTransport` could clearly host these new functions.